Glue documentation review (#1456)
spryker/spryker-docs: docs/sdk/dev/extend-the-sdk.md

> # SDK development

Author: ```suggestion
---
title: Extend the SDK
description: Find out how you can extend the Spryker SDK
template: howto-guide-template
redirect_from:
    - /docs/sdk/dev/extending-the-sdk.html
---
# SDK development
```

The SDK offers different extension points to enable third parties to contribute to the SDK without modifying it.

From simple to complex, the SDK can be extended by:

> - Providing additional tasks or settings via YAML definitions placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.yaml`. Those tasks can't introduce additional dependencies and are best suited to integrate existing tools that come with a standalone executable.

Author: ```suggestion
Providing additional [tasks](/docs/sdk/dev/task.html) or [settings](/docs/sdk/dev/settings.html) via a YAML definition placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.yaml`. Those tasks can't introduce additional dependencies and are best suited to integrating existing tools that come with a standalone executable.
```

> - Providing additional tasks, value resolvers, or settings via the PHP implementation placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.php`. Those tasks need to implement the [TaskInterface](https://github.com/spryker-sdk/sdk-contracts/blob/master/src/Entity/TaskInterface.php), and need to be exposed by providing a Symfony bundle to the Spryker SDK, such as `<path/to/spryker/sdk>/extension/<YourBundleName>/<YourBundleName>Bundle.php`, following the conventions of a [Symfony bundle](https://symfony.com/doc/current/bundles.html#creating-a-bundle). This approach is best suited for more complex tasks that don't require additional dependencies, for example validating the content of a YAML file by using Symfony validators.

Author: ```suggestion
Providing additional tasks, [value resolvers](/docs/sdk/dev/value-resolvers.html), or settings via a PHP implementation placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.php`. Those tasks need to implement the [TaskInterface](https://github.com/spryker-sdk/sdk-contracts/blob/master/src/Entity/TaskInterface.php) and need to be exposed by providing a Symfony bundle to the Spryker SDK, such as `<path/to/spryker/sdk>/extension/<YourBundleName>/<YourBundleName>Bundle.php`, following the conventions of a [Symfony bundle](https://symfony.com/doc/current/bundles.html#creating-a-bundle). This approach is best suited for more complex tasks that don't require additional dependencies, for example validating the content of a YAML file by using Symfony validators.
```

> - Providing additional tasks, value resolvers, or settings that come with additional dependencies. This approach follows the same guideline as the previous approach with the PHP implementation but requires building  your own [SDK docker image](./build.md) that includes those dependencies.

Author: ```suggestion
- Providing additional tasks, value resolvers, or settings that come with additional dependencies. This approach follows the same guideline as the previous approach with the PHP implementation but requires building  your own [SDK docker image](/docs/dev/sdk/build.html) that includes those dependencies.
```

redirect_from:
    - /docs/sdk/dev/extending-the-sdk.html
---
# SDK development

The SDK offers different extension points to enable third parties to contribute to the SDK without modifying it.

From simple to complex, the SDK can be extended by:

> Providing additional [tasks](/docs/sdk/dev/task.html) or [settings](/docs/sdk/dev/settings.html) via a YAML definition placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.yaml`. Those tasks can't introduce additional dependencies and are best suited to integrating existing tools that come with a standalone executable.

Reviewer: shouldn't this and the next paragraphs be list items, as the last one in like 16?

Author: I'm unsure what you mean here, this is 16 right?:

"- Providing additional tasks, value resolvers, or settings that come with additional dependencies. This approach follows the same guideline as the previous approach with the PHP implementation but requires building  your own [SDK docker image](/docs/sdk/dev/building-flavored-spryker-sdks.html) that includes those dependencies."

What's the difference that 14 should have and 16 has?

Reviewer: I mean you have removed "-" from paragraphs 13 and 14, which makes them plain text, and not list items. But looks like they should still be the list items, and not plain text

> Providing additional tasks, [value resolvers](/docs/sdk/dev/value-resolvers.html), or settings via a PHP implementation placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.php`. Those tasks need to implement the [TaskInterface](https://github.com/spryker-sdk/sdk-contracts/blob/master/src/Entity/TaskInterface.php) and need to be exposed by providing a Symfony bundle to the Spryker SDK, such as `<path/to/spryker/sdk>/extension/<YourBundleName>/<YourBundleName>Bundle.php`, following the conventions of a [Symfony bundle](https://symfony.com/doc/current/bundles.html#creating-a-bundle). This approach is best suited for more complex tasks that don't require additional dependencies, for example validating the content of a YAML file by using Symfony validators.

Author: ```suggestion
- Providing additional tasks, [value resolvers](/docs/sdk/dev/value-resolvers.html), or settings via a PHP implementation placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.php`. Those tasks need to implement the [TaskInterface](https://github.com/spryker-sdk/sdk-contracts/blob/master/src/Entity/TaskInterface.php) and need to be exposed by providing a Symfony bundle to the Spryker SDK, such as `<path/to/spryker/sdk>/extension/<YourBundleName>/<YourBundleName>Bundle.php`, following the conventions of a [Symfony bundle](https://symfony.com/doc/current/bundles.html#creating-a-bundle). This approach is best suited for more complex tasks that don't require additional dependencies, for example validating the content of a YAML file by using Symfony validators.
```

From simple to complex, the SDK can be extended by:

Providing additional [tasks](/docs/sdk/dev/task.html) or [settings](/docs/sdk/dev/settings.html) via a YAML definition placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.yaml`. Those tasks can't introduce additional dependencies and are best suited to integrating existing tools that come with a standalone executable.
Providing additional tasks, [value resolvers](/docs/sdk/dev/value-resolvers.html), or settings via a PHP implementation placed inside `<path/to/spryker/sdk>/extension/<YourBundleName>/Task/<taskname>.php`. Those tasks need to implement the [TaskInterface](https://github.com/spryker-sdk/sdk-contracts/blob/master/src/Entity/TaskInterface.php) and need to be exposed by providing a Symfony bundle to the Spryker SDK, such as `<path/to/spryker/sdk>/extension/<YourBundleName>/<YourBundleName>Bundle.php`, following the conventions of a [Symfony bundle](https://symfony.com/doc/current/bundles.html#creating-a-bundle). This approach is best suited for more complex tasks that don't require additional dependencies, for example validating the content of a YAML file by using Symfony validators.

> - Providing additional tasks, value resolvers, or settings that come with additional dependencies. This approach follows the same guideline as the previous approach with the PHP implementation but requires building  your own [SDK docker image](/docs/dev/sdk/build.html) that includes those dependencies.

Author: ```suggestion
- Providing additional tasks, value resolvers, or settings that come with additional dependencies. This approach follows the same guideline as the previous approach with the PHP implementation but requires building  your own [SDK docker image](/docs/sdk/dev/building-flavored-spryker-sdks.html) that includes those dependencies.
```

### Implementation via a PHP class

If a task is more than just a call to an existing tool, you can implement the task as a PHP class and register it using the Symfony service tagging feature.
This requires you to make the task a part of the Symfony bundle. To achieve this, follow these steps:

1. Create a Symfony bundle.<br>
Refer to the [official Symfony documentation](https://symfony.com/doc/current/bundles.html) for details on how to do that.

> {% info_block infoBox "Info" %}

Author: ```suggestion
{% info_block infoBox Info %}
```

4. Implement placeholders.<br>
Placeholders are resolved at runtime by using a specified value resolver.
A placeholder needs a specific name that is not used anywhere else in the command the placeholder is used for.

> {% info_block infoBox "Info" %}

Author: ```suggestion
{% info_block infoBox Info %}
```

6. Register your bundle.

If your bundle does not have dependencies that differ from the Spryker SDK, you don't need to register the bundle. Instead, place it into the `extension` directory that is a part of your SDK installation.

> For more complex bundles that require additional dependencies, follow the guidelines in [Building a flavored Spryker SDK](build.md).

Author: ```suggestion
For more complex bundles that require additional dependencies, follow the guidelines in [Building a flavored Spryker SDK](/docs/dev/sdk/build.html).
```

6. Register your bundle.

If your bundle does not have dependencies that differ from the Spryker SDK, you don't need to register the bundle. Instead, place it into the `extension` directory that is a part of your SDK installation.

> For more complex bundles that require additional dependencies, follow the guidelines in [Building a flavored Spryker SDK](/docs/dev/sdk/build.html).

Author: ```suggestion
For more complex bundles that require additional dependencies, follow the guidelines in [Building a flavored Spryker SDK](/docs/sdk/dev/building-flavored-spryker-sdks.html).
```

## 2. Add a value resolver

Most placeholders need a solution to resolve their values during runtime. This can be reading some settings and assembling a value based on the settings content, or any solution that turns a placeholder into a resolved value.

> {% info_block warningBox "" %}

Author: ```suggestion
{% info_block warningBox %}
```

## 3. Add a setting

A bundle can add more settings that value resolvers can then use to create a persistent behavior.
You can define settings in the `settings.yaml` file and add them to the SDK by calling
`spryker-sdk setting:set setting_dirs <path to your settings>`:

```yaml
settings:
  - path: string #e.g.: some_setting
    initialization_description: string #Will be used when a user is asked to provide the setting value
    strategy: string #merge or overwrite, where merge will add the value to the list and overwrite will replace it

>     init: bool #if the user should be asked for the setting value when `spryker-sdk sdk:init:sdk` or `spryker-sdk sdk:init:project` is called

Author: ```suggestion
    init: bool #if true, the user should be asked for the setting value when `spryker-sdk sdk:init:sdk` or `spryker-sdk sdk:init:project` is called.
```

    type: string #Use array for lists of values or any scalar type (string|integer|float|boolean)

>     is_project: boolean #defines if the setting is persisted across projects and initialized during `spryker-sdk sdk:init:sdk` or per project and initialized with `spryker-sdk sdk:init:project`

Author: ```suggestion
    is_project: boolean #If true, the setting is persisted across projects and initialized during `spryker-sdk sdk:init:sdk` or per project and initialized with `spryker-sdk sdk:init:project`
```
